source/base/index_set.cc

```diff
@@ -63,16 +63,32 @@ IndexSet::IndexSet (const Epetra_Map &map)
 IndexSet::IndexSet (const Epetra_Map &map)
   :
   is_compressed (true),
-  index_space_size (map.NumGlobalElements()),
+#ifndef DEAL_II_WITH_64BIT_INDICES
+  index_space_size (1+map.MaxAllGID()),
+#else
+  index_space_size (1+map.MaxAllGID64()),
+#endif
   largest_range (numbers::invalid_unsigned_int)
 {
+#ifndef DEAL_II_WITH_64BIT_INDICES
+  Assert(
+    map.MinAllGID() == 0,
+    ExcMessage("The Epetra_Map does not contain the global index 0, which "
+               "means some entries are not present on any processor."));
+#else
+  Assert(
+    map.MinAllGID64() == 0,
+    ExcMessage("The Epetra_Map does not contain the global index 0, which "
+               "means some entries are not present on any processor."));
+#endif
+
   // For a contiguous map, we do not need to go through the whole data...
   if (map.LinearMap())
     add_range(size_type(map.MinMyGID()), size_type(map.MaxMyGID()+1));
   else
     {
       const size_type n_indices = map.NumMyElements();
-      unsigned int *indices = (unsigned int *)map.MyGlobalElements();
+      unsigned int *indices = reinterpret_cast<unsigned int *>(map.MyGlobalElements());
```

Comment:
That explains my confusion -- we look at the code with its 3 32/64-bit differences :-)